src/Client/Connection.cpp

```diff
@@ -171,8 +174,26 @@ void Connection::sendHello()
     // NOTE For backward compatibility of the protocol, client cannot send its version_patch.
     writeVarUInt(client_revision, *out);
     writeStringBinary(default_database, *out);
-    writeStringBinary(user, *out);
-    writeStringBinary(password, *out);
+    /// If interserver-secret is used, one do not need password
+    /// (NOTE we do not check for DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET, since we cannot ignore inter-server secret if it was requested)
+    if (!cluster_secret.empty())
+    {
+        writeStringBinary(INTERSERVER_SECRET_MARKER, *out);
+        writeStringBinary(user /* real user as password, since user is required for hashing */, *out);
```

Comment:
Can explain a bit more what hashing you mean here?

Reply:
At first it was just one more variable for hashing, but later I find out that,
there can be the case when intial_user is empty, comment describe this:

```c
/// initial_user can be empty in case of Distributed INSERT via Buffer/Kafka,
/// i.e. when the INSERT is done with the global context (w/o user).
if (!client_info.initial_user.empty())
{
    query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
    LOG_DEBUG(log, "User (initial): {}", query_context->getUserName());
}
else
{
    query_context->setUserWithoutCheckingPassword(user_from_hello, socket().peerAddress());
    LOG_DEBUG(log, "User (from hello): {}", query_context->getUserName());
}
```

Reply:
`StorageBuffer` and `StorageKafka` both use the global context when they want to insert to a target table.
The global context is used for internal operations and it doesn't restraint access rights. That's why the user name is empty in the global context. 

I mean it seems there is no point to call `setUserWithoutCheckingPassword()` for `user_from_hello` here if `client_info.initial_user.empty()`. Because it doesn't add any extra security (you can't even check password). The only thing this call is going to do is bad: it will check that the user defined in the cluster configuration exists on the nodes and it will not allow us to remove that user from the cluster configuration. One of the reasons why I like the idea of the interserver secret is that we will able to get rid of defining the "remote" users in the cluster configuration.

Reply:
> I mean it seems there is no point to call setUserWithoutCheckingPassword() for user_from_hello here if client_info.initial_user.empty().

But StorageKafka/StorageBuffer can be then passed to the StorageDistributed (via MaterializedView), and it will not have any user, initial_user is empty and user_from_hello will be empty on the remote side, so what user should be used in this case?

Reply:
No user. If you don't call any kind of `setUser*()` function you'll get a context with no user and no restraints on access rights just like the global context. And it's ok: StorageBuffer doesn't check access rights when it inserts to a target table locally, so it's fine not to check access rights when it works on remote nodes (as long as the interserver secret protects our connections).

Reply:
Ok

Reply:
Done